service/storage_service.cc

```diff
@@ -1126,6 +1126,53 @@ class topology_coordinator {
         }
     }
 
+    future<> enable_features(group0_guard guard, std::set<sstring> features_to_enable) {
+        if (!_topo_sm._topology.new_nodes.empty()) {
+            // If there are some join requests present, then we aren't safe
+            // to enable features now. We don't include those nodes' features
+            // when calculating features to enable, and we are not sure whether
+            // we already promised to them that we will join them, so it's not
+            // safe to enable features now.
+            on_internal_error(slogger,
+                    "topology coordinator attempted to enable features even though there are joining nodes");
+        }
+
+        // If we are here, then we noticed that all nodes support some features
+        // that are not enabled yet. Perform a global barrier to make sure that:
+        //
+        // 1. All nodes saw (and persisted) a view of the system.topology table
+        //    that is equal to what the topology coordinator sees (or newer,
+        //    but in that case updating the topology state will fail),
+        // 2. None of the nodes is restarting at the moment and trying to
+        //    update its corresponding `supported_features` column (that's why
+        //    we use `barrier_normal` instead of regular `barrier`).
+        //
+        // After we get a successful confirmation from each node, we have
+        // a guarantee that they won't attempt to revoke support for those
+        // features. That's because we do not allow nodes to boot without
+        // a feature that is supported by all nodes in the cluster, even if
+        // the feature is not enabled yet.
+        {
+            auto f = co_await coroutine::as_future(exec_global_command(std::move(guard),
+                    raft_topology_cmd{raft_topology_cmd::command::barrier_normal},
+                    {_raft.id()}));
+            if (f.failed()) {
+                slogger.error("raft topology: enable_features, "
```

Comment:
Shouldn't we be throttling this step upon a failure?

Reply:
Yes, sounds reasonable.

There are some other places in the code which react to failures in the same way (e.g. handling `topology::transition_state::commit_cdc_generation`), so we should probably handle them as well. I'm thinking about introducing `enum class handle_result { no_work, had_work, need_retry }` that would be propagated down to the `topology_coordinator::run()` instead of the current `bool` method and handled appropriately there.